lgc/include/lgc/util/CacheHashNoteUtils.h

```diff
@@ -0,0 +1,98 @@
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  CacheHashNoteUtils.h
+ * @brief LLPC header file: contains the definition of LLPC utility function
+ * addHashSectionToElf.
+ *
+ * @details The function addHashSectionToElf adds a note section
+ *          "llpc_cache_hash" (see CacheHashNoteName) in the given ELF to
+ *          keep the hash code for the cache, which will be used by cache
+ *          creator to identify the mapping between the hash and the shader
+ *          ELF.
+ *
+ * @sa llpc/docs/CacheHashNoteUtils.md
+ ***********************************************************************************************************************
+ */
+
+#pragma once
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace lgc {
+
+constexpr char CacheHashNoteName[] = "llpc_cache_hash"; // Cache hash note name for .note record.
+constexpr char LlpcVersionNoteName[] = "llpc_version";  // LLPC version note name for .note record.
+
+// =====================================================================================================================
+// Adds a note section "llpc_cache_hash" (see CacheHashNoteName) in the given
+// ELF to keep the hash code for the cache.
+//
+// ELF layouts before/after adding new ".note" section and its section header
+// for cache hash:
+//
+// |-------------|        |-------------|
+// | ELF header  |        | ELF header  |
+// |-------------|        |-------------|
+// | Sections    |        | Sections    |
+// | ...         |        | ...         |
+// |-------------|  ==>   |-------------|
+// | Section     |        | Section     |
+// | headers     |        | headers     |
+// | ...         |        | ...         |
+// |-------------|        |-------------|
+// | Sections    |        | New section |
+// | ...         |---|    | header for  |
+// |-------------|   |    | hash note   |    This new section header's
+//                   |    | section    +|--| offset must be the new
+//    Remaining      |    |-------------|  | note section for cache hash
+//    sections after |    | New note    |  |
+//    section header |    | section for |<-/
+//    table must be  |    | cache hash  |
+//    shifted        |    |-------------|
+//                   \--->| Sections    |
+//                        | ...         |
+//                        |-------------|
+//
+// The new note section for the cache hash will contain two note entries:
```

Comment:
So this adds a new .note section? Is it legal to have more than one .note section?

Reply:
As far as I understand, the [spec](https://man7.org/linux/man-pages/man5/elf.5.html) (or http://www.skyfree.org/linux/references/ELF_Format.pdf) does not specify any limitation about the number of sections with the same type. Similar other sections like relocatable sections and symbol table sections, I think it is fine to add multiple `.note` sections.

Reply:
I think you might be right.  You can have multiple note section but they should probably have different names.  An elf file I happen to have on hand has two note sections:

```
$ llvm-objdump -mcpu=gfx900 t.elf -h | less  

t.elf:  file format ELF64-amdgpu

Sections:
Idx Name          Size     VMA          Type
  0               00000000 0000000000000000 
  1 .strtab       00000078 0000000000000000 
  2 .text         000001a0 0000000000000000 TEXT 
  3 .rel.text     00000040 0000000000000000 
  **4 .note         00000348 0000000000000000** 
  5 .AMDGPU.disasm 00000a67 0000000000000000 
  **6 .note.GNU-stack 00000000 0000000000000000** 
  7 .symtab       00000090 0000000000000000 
```